python/pyiceberg/typedef.py

```diff
@@ -79,6 +84,10 @@ def __missing__(self, key: K) -> V:
 class StructProtocol(Protocol):  # pragma: no cover
     """A generic protocol used by accessors to get and set at positions of an object"""
 
+    @abstractmethod
+    def set_record_schema(self, record_schema: StructType) -> None:
```

Comment:
One thing that I didn't think about when we were discussing where to put this method is that this is can be called multiple times if we don't pass the schema to the constructor, allowing callers to change the record schema. We assume that it will only be set once in the readers, but this makes the API quite strange. I can read with one schema then set a different schema, and then the behavior is unclear.

Because of that, I'd prefer to move this method to our `PydanticStruct` class instead of putting it here. That way we control all of the implementations and it doesn't matter. And we will have the ability to change it later if needed.

Reply:
I've merged the two records into one now. Changing the schema makes things more complicated. For example, then we should also unset fields, which we can do, but we make it more complex. For the sake of keeping things simple, we could limit the reuse of the record within the StructReader, and this way it always reads the same schema.